Test suite for transfer edges indexing

adzialocha opened the pull request:
This test suite should automatically test against the new indexing algorithm creating the `edges.json` file based on blockchain events and graph data. Related PR: https://github.com/CirclesUBI/circles-api/pull/24

**Basic test setup:**

* [x] Run mocked Blockchain via `ganache-core` (example: https://hackernoon.com/minimal-solidity-contract-testing-with-ganache-and-jest-f735547d9643)
* [x] Deploy `Hub` and Gnosis `Safe` and `ProxyFactory` contracts (similar to https://github.com/CirclesUBI/circles-contracts/tree/master/migrations)
* [x] Write some helper methods to create a new `Safe` and `Token`
* [ ] Create test cases based on https://hackmd.io/G2txlqZaTciU5rTx-wA-gw specification, https://gist.github.com/adzialocha/000845f42a4cd951ef64feb0f47d2864 and https://github.com/CirclesUBI/circles-contracts/blob/trust-graph-testing/trustGraphTesting.js
* [ ] Call regarding indexing methods in `circles-api`
* [ ] Read `edges.json` file in `pathfinder` folder after indexing and compare contents with test vectors (all in .json format)
* [ ] Maybe poll contents for a certain amount of time to reflect async behaviour (eg. `edges.json` should reflect test vector contents within 10 seconds etc.) (similar to https://github.com/CirclesUBI/circles-core/blob/main/test/helpers/loop.js#L24)

**Nice to have:**

* [ ] Mock ioredis, or use a real test redis instance (https://github.com/CirclesUBI/circles-api/issues/67)
* [ ] Mock graph responses according to test cases (similar to https://github.com/CirclesUBI/circles-api/blob/main/test/utils/mocks.js) (https://github.com/CirclesUBI/circles-api/issues/68)

ana0 commented on src/connections/blockchain.js:
```diff
@@ -14,10 +14,13 @@ const fuelProvider = new FuelProvider(
 )
 
-const web3 = new Web3(fuelProvider.start());
+console.log(fuelProvider.start)
```
No console

ana0 commented on src/connections/blockchain.js:
```diff
@@ -14,10 +14,13 @@ const fuelProvider = new FuelProvider(
 
-const web3 = new Web3(fuelProvider.start());
+console.log(fuelProvider.start)
+let provider = fuelProvider.start()
```
Don't think this gets redefined?

ana0 commented on src/controllers/relayerController.js:
```diff
@@ -0,0 +1,19 @@
+
+async function signup (req, res) {
+  try {
+    let receipt = await HubContract.methods.signup(req.body.address, req.body.name)
```
Think this isn't redefined either

ana0 commented on src/connections/blockchain.js:
```diff
@@ -1,10 +1,10 @@
 const fuelProvider = new FuelProvider(
   {
-    privateKey: process.env.PRIVATE_KEY,
+    privateKey: process.env.PRIVATE_KEY.slice(2),
```
Is this slicing off a '0x' or somethin?

ana0 commented:
These are all v small, just gonna fix and merge

louilinn committed:
Setup ganache for edges tests

llunaCreixent committed:
Set up Hub for edges tests

llunaCreixent committed:
Fix format in edges test with ESLint + Prettier.

llunaCreixent committed:
Create Safes in edges test. Create utils/safes.js

adzialocha commented on test/utils/safes.js:
```diff
@@ -0,0 +1,58 @@
+// Helper function used in tests to create Safe instances associated to the
+// given ganache accounts. The account 0 is used as the admin account
+// and it doesn't have a Safe.
+export async function createSafes(accounts) {
```
This looks really cool! I suggest having two method arguments, one for the funder account and the others for an array of the actual accounts to be added to the created safes. Then the use of `accounts[0]` is more explicit.

llunaCreixent replied:
Done ! 😺

llunaCreixent committed:
Added one argument to the createSafes() function in the edges tests.

llunaCreixent commented:
Should we implement the signup process of the accounts? (Like in https://github.com/CirclesUBI/circles-contracts/blob/trust-graph-testing/trustGraphTesting.js)

adzialocha commented:
> Should we implement the signup process of the accounts? (Like in https://github.com/CirclesUBI/circles-contracts/blob/trust-graph-testing/trustGraphTesting.js)

Yes, we will need this as well to be part of the trust network 👍

louilinn committed:
Execute transactions for account signup in edges tests

llunaCreixent committed:
Signup the Safes and create tokens in edges tests.

louilinn committed:
Add test environment variable and using NODE_ENV in edges.json filename

louilinn committed:
Using edges-test.json for testing - work in progress

llunaCreixent commented:
Should we export the `rebuildTrustNetwork()` in `worker.js` conditioned to the `ENV_NODE` value?

adzialocha commented:
> Should we export the `rebuildTrustNetwork()` in `worker.js` conditioned to the `ENV_NODE` value?

The export could be part of all environments, it does not need to be exported only for the `test` case

adzialocha commented:
So, I merged `logging` into `main` and rebased this branch against it, from now on we can test against the new and right system :sparkles: 

Also, I've moved the `WORKER_TESTCASES.md` into this gist: https://gist.github.com/adzialocha/000845f42a4cd951ef64feb0f47d2864 (link also in PR description above).

llunaCreixent commented:
We use the [web3 utility from the tests directory](https://github.com/CirclesUBI/circles-api/blob/6b0b0707192ffb436b19904c5df73b3aeb2df301/test/edges-e2e.test.js#L4), but we also prepared the web3 service in the `src` directory to look for the [provider in the web3 from the test utilities](https://github.com/CirclesUBI/circles-api/blob/6b0b0707192ffb436b19904c5df73b3aeb2df301/src/services/web3.js#L6). 
It's redundant as we are not using the `src/services/web3` file in our tests.

adzialocha commented:
> We use the [web3 utility from the tests directory](https://github.com/CirclesUBI/circles-api/blob/6b0b0707192ffb436b19904c5df73b3aeb2df301/test/edges-e2e.test.js#L4), but we also prepared the web3 service in the `src` directory to look for the [provider in the web3 from the test utilities](https://github.com/CirclesUBI/circles-api/blob/6b0b0707192ffb436b19904c5df73b3aeb2df301/src/services/web3.js#L6).
> It's redundant as we are not using the `src/services/web3` file in our tests.

I'm not sure what exactly is redundant here because both methods to connect to the blockchain are required (one for running on production, one for running the tests), but I agree we can consider moving them into the same file, though I also like the separation because its clear that one way is only related to tests, also the test web3 provider uses `ganache-core` and if we would move this into `services/web3` we would need to use it also during production. This would introduce an unnecessary dependency.

llunaCreixent committed:
Call worker to listen to events and read/delete edges.json file in edges tests

llunaCreixent committed:
Delete unnecessary import

llunaCreixent commented:
I couldn't find a way to mock the ioredis; so far we need to have the redis instance running locally (e.g. in docker).

llunaCreixent committed:
Create test cases, step 1, no edges created in edges-test.json

llunaCreixent commented:
Step 1 of the [Trust graph test cases](https://gist.github.com/adzialocha/000845f42a4cd951ef64feb0f47d2864) working 🎈 ! I'm using the docker environment with `make up EXPOSE_PORTS=1`.

llunaCreixent committed:
Don't delete edges.json file after edges tests

llunaCreixent commented:
Trying to run the original code of the worker in an async process, but I get an error:
`SyntaxError: Cannot use import statement outside a module`.

The asynchronous process creation code used:
```js
const worker = exec('cross-env NODE_ENV=test node --input-type=module ./src/worker.js ', (error, stdout, stderr) => {
      if (error || stderr) {
        console.log(error);
        console.error(`Worker process failed`);
        return;
      }
      console.log(stdout);
      console.log("Worker runs");
    });
```

llunaCreixent commented:
> Trying to run the original code of the worker in an async process, but I get an error:
> `SyntaxError: Cannot use import statement outside a module`.
> 
> The asynchronous process creation code used:
> 
> ```js
> const worker = exec('cross-env NODE_ENV=test node --input-type=module ./src/worker.js ', (error, stdout, stderr) => {
>       if (error || stderr) {
>         console.log(error);
>         console.error(`Worker process failed`);
>         return;
>       }
>       console.log(stdout);
>       console.log("Worker runs");
>     });
> ```

Solved: first `npm run build` and use the worker in `build/`.

llunaCreixent committed:
Run  worker code in a child process

llunaCreixent commented:
The code of the worker runs in a child process. But now the mocking the graph does not work. The `edges-e2e` tests need the docker environment (with `make up EXPOSE_PORTS=1` and `make subgraph`).
